sycamore/evaluation/pipeline.py

```diff
@@ -0,0 +1,115 @@
+                                }
+                            }
+                        ],
+                        "filter": [{"exists": {"field": "parent_id"}}],
+                    }
+                },
+                "ext": {
+                    "generative_qa_parameters": {
+                        "llm_question": doc["question"],
+                        "context_size": self._os_config["context_window"],
+                        "llm_model": self._os_config["llm"],
+                    }
+                },
+                "size": self._os_config["size"],
+            }
+        else:
+            query["query"] = {
+                "_source": {"excludes": ["embedding"]},
+                "query": {
+                    "bool": {
+                        "must": [
+                            {
+                                "hybrid": {
+                                    "queries": [
+                                        {"match": {"text_representation": doc["question"]}},
+                                        {
+                                            "neural": {
+                                                "embedding": {
+                                                    "query_text": doc["question"],
+                                                    "model_id": self._os_config["embedding_model_id"],
+                                                    "k": self._os_config["neural_search_k"],
+                                                }
+                                            }
+                                        },
+                                    ]
+                                }
+                            }
+                        ],
+                        "filter": [{"exists": {"field": "parent_id"}}],
+                    }
+                },
+                "size": self._os_config["size"],
+            }
+        return query
+
+    def _evaluate_queries(self, query: Document) -> Document:
+        result = EvaluationDataPoint(query)
+        metrics = {}
+        for metric in self._metrics:
+            metrics[metric.metric_name()] = metric.evaluate(result)
+        result["metrics"] = metrics
+        return result
+
+    def execute(self, input_dataset: DocSet) -> Tuple[DocSet, DocSet]:
+        # 1. convert input dataset to opensearch queries [EvaluationDataPoint -> OpenSearchQuery]
+        opensearch_queries = input_dataset.map(self._build_opensearch_query)
+
+        # 2. execute opensearch queries [OpenSearchQuery -> OpenSearchQueryResult]
+        opensearch_results = opensearch_queries.query(query_executor=self._query_executor)
+
+        # 3. query level metrics [OpenSearchQueryResult -> EvaluatedEvaluationDataPoint]
+        query_level_metrics = opensearch_results.map(self._evaluate_queries)
+
+        # 4. aggregation metrics [[EvaluatedEvaluationDataPoint] -> EvaluatedQASummary]
+        aggregated_metrics = query_level_metrics.map(lambda x: x)
```

Comment:
maybe make the aggregation function a method? So I can subclass this pipeline and do fancy custom stuff

Reply:
oh this is a placeholder, Ben is going to help setup the implementation with some other stuff he's done